### PR TITLE
Add mlflow converter for video-multi-object-tracking task

### PR DESCRIPTION
**Pull request by yiyoulin**
run with latest changes: https://ml.azure.com/experiments/id/a7c4f52c-fdbe-440a-a2c0-1ccde8104c82/runs/brave_square_fmcdgdcnpk?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/vidani_rg/providers/Microsoft.MachineLearningServices/workspaces/vidani_wrk&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

### CODE REVIEW

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                self._model = init_model(self._config, model_weights_path)
+
+                print("Model loaded successfully")
+            except Exception:
+                print("Failed to load the the model.")
+                raise
+
+        else:
+            raise ValueError(f"invalid task type {self._task_type}."
+                             f"Supported tasks: video-multi-object-tracking")
```
Comment (PhaniShekhar) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:128]: Please use constant for task here as well

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+    def predict(self, context: mlflow.pyfunc.PythonModelContext, input_data: pd.DataFrame) -> pd.DataFrame:
+        """Perform inference on the input data.
+
+        :param context: MLflow context containing artifacts that the model can use for inference
+        :type context: mlflow.pyfunc.PythonModelContext
+        :param input_data: Input images for prediction
```
Comment (PhaniShekhar) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:135]: The description for input_data needs to change. Also the type description needs to be related to "video" column

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        """This method is called when the python model wrapper is initialized.
+
+        :param task_type: Task type used in training.
+        :type task_type: str
+        """
+        super().__init__()
+        self._model = None
+        self._task_type = task_type
+        self._inference_detector = None
+        self.video_reader = None
```
Comment (PhaniShekhar) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:79]: Nit: use self._video_reader ?

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        """Parse the output of inference_mot() to a list of dictionaries.
+        param result: output of inference_mot()
+        type result: tuple
+        param frame_id: frame id of the current frame
+        type frame_id: int
+        param video_url: url of the video
+        type video_url: str
+        return: dictionary of current image predictions
+        """
+        print(result.keys())
```
Comment (PhaniShekhar) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py]: Is this print statement required  ?

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        det_bboxes = result[MOTLiterals.DET_BBOXES]
+        track_bboxes = result[MOTLiterals.TRACK_BBOXES]
+
+        curimage_preds = {MOTLiterals.DET_BBOXES: [], MOTLiterals.TRACK_BBOXES: [],
+                          MOTLiterals.FRAME_ID: frame_id, MOTLiterals.VIDEO_URL: video_url}
+        for label, bboxes in enumerate(det_bboxes):
+            for bbox in bboxes:
```
Comment (PhaniShekhar) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:172]: In the corresponding function in mmtrack_utils.py, we have a check for if label >= 0. Is that not required here ?
Also, the inner for-loop "for bbox in bboxes:" is missing in mmtrack_utils.py. Is there any difference in det_bboxes format between the two

**Review by PhaniShekhar — CHANGES_REQUESTED** [review]

File: assets/training/model_management/components/convert_model_to_mlflow/spec.yaml [hunk 6]
```diff
@@ -44,6 +44,7 @@ inputs:
       - Huggingface
       - MMLab
       - llava
+      - MMTrack
```
[the remark this answers is not in the document]
Comment (PhaniShekhar) [reply assets/training/model_management/components/convert_model_to_mlflow/spec.yaml]: Thanks for updating this. Resolving.

File: assets/training/model_management/src/azureml/model/mgmt/config.py [hunk 7]
```diff
@@ -30,6 +30,7 @@ class ModelFramework(_CustomEnum):
     HUGGINGFACE = "Huggingface"
     MMLAB = "MMLab"
     LLAVA = "llava"
+    MMTRACK = "MMTrack"
```
[the remark this answers is not in the document]
Comment (PhaniShekhar) [reply assets/training/model_management/src/azureml/model/mgmt/config.py]: Thanks for updating this. Resolving.

**Comment by yiyoulin** [comment]
latest run with restructured code: framework MmTrack get combined into MmLab. 

https://ml.azure.com/experiments/id/a7c4f52c-fdbe-440a-a2c0-1ccde8104c82/runs/gray_quill_s675vkjf0x?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/vidani_rg/providers/Microsoft.MachineLearningServices/workspaces/vidani_wrk&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

**Review by vizhur — CHANGES_REQUESTED** [review]
You should probably update environment to the version that has no past due vulnerabilities

File: assets/training/model_management/src/azureml/model/mgmt/processors/factory.py [hunk 8]
```diff
@@ -15,6 +15,7 @@
 )
 from azureml.model.mgmt.processors.pyfunc.config import (
     MMLabDetectionTasks,
+    MMTrackVideoTasks,
```
[the remark this answers is not in the document]
Comment (yiyoulin) [reply assets/training/model_management/src/azureml/model/mgmt/processors/factory.py:18]: changed :) thanks for the advice!

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+import sys
+from typing import Dict, List
+
+import mlflow
+import pandas as pd
+import torch
+from config import MMDetLiterals, MLflowSchemaLiterals, MOTLiterals
+
+
+def _is_valid_url(text: str) -> bool:
```
[the remark this answers is not in the document]
Comment (yiyoulin) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:17]: thanks, changed :) previously I think of putting all the functions inside the same file to avoid since there's only one function needs to be imported. but you are right, I can totally re-use it.

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    """
+    video = vid[0]
+    if isinstance(video, str):
+        if _is_valid_url(video):
+            return video
+    raise ValueError("Video received is not in valid format. Expected format is url string.")
+
+
+class VideosTrackingMLflowModelWrapper(mlflow.pyfunc.PythonModel):
+    """MLFlow model wrapper for AutoML for Images models."""
```
[the remark this answers is not in the document]
Comment (yiyoulin) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:64]: changed  ;)

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py (hunk 3 above)
Comment (yiyoulin) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:79]: changed :)

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py (hunk 2 above)
Comment (yiyoulin) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:135]: changed :)

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py (hunk 5 above)
Comment (yiyoulin) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:172]: yes, the output from the two are different. for this one, there are only two: "track_bboxes", "det_bboxes", but in the mlflow wrapper there are four: "track_bboxes", "track_labels", "det_bboxes", "det_labels". In short, it's organized version. So we will have to test if the label is >=0. here we can just look through the ndarray, and parse the output.

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py (hunk 4 above)
Comment (yiyoulin) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py]: deleted :)

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py (hunk 1 above)
Comment (yiyoulin) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:128]: changed :)

**Comment by yiyoulin** [comment]
> You should probably update environment to the version that has no past due vulnerabilities

Hi @vizhur , could you specify which of the dep is past due vulnerability? I will test individually if I could upgrade the package. 

The requirements in conda.yaml are from dependency list of mmtrack==0.14.0. there's problem installing them directly in some of the cuda image, so I will have to list them out here. mmtrack==0.14.0 is the latest version of mmtrack we could use now.

if you are referring to any other dependencies, pls let me know as well, I could either resolve in this PR or in a subsequent pr. thank you!

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (PhaniShekhar) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py]: Thanks for updating

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py (hunk 5 above)
Comment (PhaniShekhar) [reply assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/vision/track_predict.py:148]: Thanks for the clarification :)

**Review by PhaniShekhar — APPROVED** [review]

**Comment by yiyoulin** [comment]
> You should probably update environment to the version that has no past due vulnerabilities
I updated transformers==4.33.3, torch & torchvision. let me know if it's sufficient.

**Comment by yiyoulin** [comment]
most recent run: https://ml.azure.com/experiments/id/a7c4f52c-fdbe-440a-a2c0-1ccde8104c82/runs/teal_bread_dc6p0l4gcg?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/vidani_rg/providers/Microsoft.MachineLearningServices/workspaces/vidani_wrk&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

with updated dependencies (to resolve vulnerability):
  - torch==1.13.1
  - torchvision==0.14.1
  - transformers==4.33.1

**Review by PhaniShekhar — APPROVED** [review]
